akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala

```diff
@@ -120,35 +120,81 @@ private[http] object Websocket {
-        new RouteLogic[Either[BypassEvent, MessagePart]] {
-          def initialState: State[_] = State(DemandFromAny(s)) { (ctx, out, ev) ⇒
-            ev match {
-              case Left(_) ⇒
-                State(DemandFrom(s.out0)) { (ctx, _, ev) ⇒ // FIXME: #17004
-                  ctx.emit(s.out0)(ev.left.get)
-                  initialState
-                }
-              case Right(_) ⇒
-                State(DemandFrom(s.out1)) { (ctx, _, ev) ⇒
-                  ctx.emit(s.out1)(ev.right.get)
-                  initialState
-                }
+    def bypassRouter = new GraphStage[FanOutShape2[Output, BypassEvent, MessagePart]] {
+      private val in = Inlet[Output]("in")
+      private val bypass = Outlet[BypassEvent]("bypass-out")
+      private val user = Outlet[MessagePart]("message-out")
+
+      def shape = new FanOutShape2(in, bypass, user)
+      def createLogic = new GraphStageLogic {
+
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            grab(in) match {
+              case b: BypassEvent with MessagePart ⇒ emit(bypass, b, emit(user, b, pull(in)))
+              case b: BypassEvent                  ⇒ emit(bypass, b, pull(in))
+              case m: MessagePart                  ⇒ emit(user, m, pull(in))
             }
           }
+        })
+
+        setHandler(bypass, EagerTerminateOutput)
+        setHandler(user, IgnoreTerminateOutput)
+
+        override def preStart(): Unit = {
+          super.preStart()
+          pull(in)
+        }
+
+        /*
+         * Everything below to be pushed into GraphStageLogic for reuse
+         */
+
+        private object EagerTerminateOutput extends OutHandler {
+          override def onPull(): Unit = ()
+        }
+
+        private object IgnoreTerminateOutput extends OutHandler {
+          override def onPull(): Unit = ()
+          override def onDownstreamFinish(): Unit = ()
+        }
+
+        private def emit[T](out: Outlet[T], elem: T, andThen: ⇒ Unit): Unit = {
+          if (isAvailable(out)) {
+            push(out, elem)
+            andThen
+          } else {
+            setHandler(out, new Emitting(out, elem, getHandler(out), () ⇒ andThen))
+          }
+        }
 
-          override def initialCompletionHandling: CompletionHandling = super.initialCompletionHandling.copy(
-            onDownstreamFinish = { (ctx, out) ⇒
-              if (out == s.out0) ctx.finish()
-              SameState
-            })
+        private def emit[T](out: Outlet[T], elem: T): Unit = {
+          if (isAvailable(out)) {
+            push(out, elem)
+          } else {
+            setHandler(out, new Emitting(out, elem, getHandler(out), DoNothing))
+          }
         }
+
+        private class Emitting[T](out: Outlet[T], elem: T, next: OutHandler, andThen: () ⇒ Unit) extends OutHandler {
+          if (getHandler(out).isInstanceOf[Emitting[_]])
```

Comment:
btw, please install a default handler to each stage, because currently the handler Arrays are null. Simply filling them with a constant InvalidInHandler and InvalidOutHandler that just throw on every event "no handler is installed" could be helpful.